src/interp.rs

```diff
@@ -19,153 +19,483 @@
-                let f = j % factor;
+        pub fn get_factor(&self) -> usize {
+            self.factor
+        }
 
-                let f = &mut filter[f];
-                f.push((c, j / factor));
+        pub fn reset(&mut self) {
+            // TODO: Use slice::fill() once stabilized
+            for v in &mut *self.z {
+                *v = 0.0;
             }
+            self.zi = 0;
         }
 
-        Interp {
-            factor,
-            taps,
-            channels,
-            delay,
-            filter: filter
-                .into_iter()
-                .map(|f| Filter {
-                    coeff: f.into_boxed_slice(),
-                })
-                .collect(),
-            z: z.into_boxed_slice(),
-            zi: 0,
+        pub fn process(&mut self, src: &[f32], dst: &mut [f32]) {
+            assert!(src.len().checked_mul(self.factor) == Some(dst.len()));
+            assert!(self.z.len() == self.delay * self.channels as usize);
+            assert!(self.filter.len() == self.factor);
+            assert!(self.zi < self.delay);
+
+            if src.is_empty() {
+                return;
+            }
+
+            let frames = src.len() / self.channels as usize;
+
+            for (src, (dst, z)) in src.chunks_exact(frames).zip(
+                dst.chunks_exact_mut(frames * self.factor)
+                    .zip(self.z.chunks_exact_mut(self.delay)),
+            ) {
+                let mut zi = self.zi;
+
+                for (src, dst) in src.iter().zip(dst.chunks_exact_mut(self.factor)) {
+                    // Add sample to delay buffer
+                    //
+                    // TODO Ringbuffer without bounds checks for z/zi
+                    //
+                    // Safety: zi is checked to be between 0 and self.delay
+                    *unsafe { z.get_unchecked_mut(zi) } = *src;
+
+                    // Apply coefficients
+                    for (filter, dst) in self.filter.iter().zip(dst.iter_mut()) {
+                        let mut acc = 0.0;
+                        for (c, index) in &*filter.coeff {
+                            let mut i = zi as i32 - *index as i32;
+                            if i < 0 {
+                                i += self.delay as i32;
+                            }
+
+                            // Safety: zi is checked to be between 0 and self.delay
+                            acc += *unsafe { z.get_unchecked(i as usize) } as f64 * c;
+                        }
+
+                        *dst = acc as f32;
+                    }
+
+                    zi += 1;
+                    if zi == self.delay {
+                        zi = 0;
+                    }
+                }
+            }
+
+            self.zi = (self.zi + frames) % self.delay;
         }
     }
+}
 
-    pub fn get_factor(&self) -> usize {
-        self.factor
+/// A trait to be generic over number of channels in a of frame
+trait Frame: Sized + Copy {
```

Comment:
Quick thought here; these traits also exist in the `sample` crate. It might be a reasonable dependency.

Reply:
Also in [`dasp`](https://github.com/rustaudio/dasp)

Not sure if this small trait here makes sense to pull in a separate dependency. Let's put a FIXME comment here and think about that some other time :)

Reply:
Neat! Had not noticed the developmen on sample->dasp. Then I suppose a dep on only dasp_frame would suffice (with only dasp_sample as transitive dep). But yes, could be for future consideration.

Reply:
`dasp_frame` as dependency (and `dasp_sample` transitively) seems OK to me. If you want, update it to that crate now or just add a comment about that :)

Reply:
I will add a comment for now. The mul_add operation is not covered in dasp, and if using dasp, it should be evaluated in context with the `Samples` trait (how they related/differ/interact)